src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelperFactory.cs

```diff
@@ -43,7 +45,18 @@ public IUrlHelper GetUrlHelper(ActionContext context)
                 return (IUrlHelper)value;
             }
 
-            var urlHelper = new UrlHelper(context);
+            IUrlHelper urlHelper;
+            var endpointFeature = httpContext.Features?.Get<IEndpointFeature>();
+            if (endpointFeature != null && endpointFeature.Endpoint != null)
+            {
+                var linkGenerator = httpContext.RequestServices.GetRequiredService<ILinkGenerator>();
+                urlHelper = new UrlHelper(context, linkGenerator);
+            }
+            else
+            {
+                urlHelper = new UrlHelper(context);
+            }
+
```

Comment:
This mostly seems fine. Note that we have more flexibility than this, I don't think we really expect some requests to be dispatcher and some not.